### ISSUE TITLE
GAN Memory with No Forgetting

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#141]
## TL; DR

- GANs can be adapted to new domain with learnable weight modulations. (similar to AdaIN, but applied to model parameters)
- Freeze source GAN and train/store style-modulation parameters for pseudo-rehearsal.

## Motivation

- Rehearsal-based CL methods suffer from memory scalability and privacy issue. (as they *keep* data samples throughout the process)
- Pseudo-rehearsal methods with generative replays also suffer from scalability issue, as the generative performance drops with longer task sequences.

## Previous Work

- Rehearsal-based
- Pseudo-rehearsal: MeRGAN, CLOGAN, DGMw
- Feature Modulation: FiLM (Feature-wise Linear Modulation), AdaFM (Adaptive Filter Modulation)

### FiLM

- Apply linear modulation to the latent features (as in AdaIN)

![image](https://user-images.githubusercontent.com/46308407/147517244-c4a95535-d9e7-4f6a-a3a0-984e80ed0e3d.png)

- h, \hat{h} are activation and the modulated latent feature, respectively. 

### AdaFM

- Apply modulation to convolution kernels (filters).
- Conceptually similar to modulated-Conv in StyleGAN2.

![image](https://user-images.githubusercontent.com/46308407/147517917-b8fb9784-e9f1-4695-acdd-0aafa29eeab8.png)

### Revelation on GAN adaptation

![image](https://user-images.githubusercontent.com/46308407/147518299-abcc2474-8d25-4814-8828-3135e150cf57.png)

- GANs trained on source dataset can be adapted to perceptually distant target domain with simple weight modulation.
- The source GAN parameters are frozen, and only the style modulation parameters are updated.

- FC layer
![image](https://user-images.githubusercontent.com/46308407/147517954-06d15968-017b-4fff-a05f-59f1afc87179.png)

- Conv layer
![image](https://user-images.githubusercontent.com/46308407/147517966-bfe57ff6-0936-436f-89f0-775838c2b10c.png)

- This is equivalent to applying batch normalization to the model weights (FC and Conv)

### GAN Memory

![image](https://user-images.githubusercontent.com/46308407/147518054-f1506b49-ccb4-4d60-b634-71f4135666ee.png)
![image](https://user-images.githubusercontent.com/46308407/147518397-1f48ee84-ad84-4dad-acb6-a24bc0512cf9.png)

- Scale, Shift and Bias are 3 groups of modulation parameters. They are responsible for different aspects of generation.
- Style modulation parameters from different blocks (vertically) control different hierarchical semantics. (coarse - fine)
- Normalization matters. Applying linear modulation without normalization results in inferior adaptation performance.
- Smooth interpolation and conditional synthesis.

### Compression

![image](https://user-images.githubusercontent.com/46308407/147518381-59b51283-0ee4-4b44-a777-7093664d102a.png)

- Scale and Shift modulation parameters for Conv are typically huge in size.
- Observing the singular values of the learned modulation matrices, PCA yields good trade-off between memory and performance.
- Low rank matrix factorization can also be applied on top of *shared knowledge base*.
- Left/Right knowledge base are shared, and task-specific parameter updates are made only to lambda(diagonal) and E.
- E matrix is norm-regularized to enforce low-rank property.

### Generation and Continual Learning

![image](https://user-images.githubusercontent.com/46308407/147518511-e99b266c-3483-4e3a-8557-f7c99e559434.png)

### Compression

![image](https://user-images.githubusercontent.com/46308407/147518533-9b3abc12-7b9f-491d-b296-83af5a4bd150.png)

## Comments

- Interesting idea to train weight modulation parameters, instead of the model weights.
- Similar concept of learnable masks (on weights) is widely employed (ref: Adversarial Generation of Continual Image, CVPR'21)
- Model compression is somewhat naive but straightforward and looks quite effective.
- However, quite skeptical about the scalability of proposed style-modulation approach. The demonstration is on low resolution images (CelebA, CIFAR-10), and I don't see applications stemming from this work in GAN adaptation literature much. Citation is moderate as well.
- More than that, if the adaptation quality was that great, the authors probably (personal opinion) needed not claim contribution for Continual Learning. In other words, the synthesis quality is rather *moderate*, but effective enough for pseudo-rehearsal based continual learning, in that the source parameters can be kept frozen.
- The memory footprint for storing style modulation parameters is not at all negligible, as they mainly apply element-wise operations. In fact, even more parameters have to be stored compared to keeping a separate model checkpoint. (scale, shift and bias)
- If the target domain lack data samples (i.e., few-shot like setting), this approach can be effective in that the source domain parameters are fixed (distance collapse can be mitigated to certain extent).
